### ISSUE TITLE
get HTTP ERROR 413 return

### ISSUE THREAD

**Issue by bigfoxtail** [mitmproxy/mitmproxy#4231]
mitmproxy version: master newest
python version: 3.8.5
OS: windows 10
browser: Chrome 85
use mitmproxy browser `https://ww1.sinaimg.cn/mw600/5f789f8fgy1gjnnwi718sj20s50funk3.jpg` ,get return 413.
i find,
commit `828ba0c2e79d8c54806a1c9eefb6007abc8aabc0` is normal,
commit `c0f62cc559ae4f0778d6079a2b31d2ce0309dda6` is return ERROR 413.

**Comment by bigfoxtail** [comment]
> We just had this in #4221 and it looked like it's fixed in master (@Kriechi). But I can confirm that HEAD is broken. Interestingly it does work for the URL in #4221.
> 
> So it looks like some of the HTTP/2 code changed and made it compatible with one server (https://g.alicdn.com/kg/kmd-adapter/0.0.6/index.js) while breaking compatibility with the other (https://ww1.sinaimg.cn/mw600/5f789f8fgy1gjnnwi718sj20s50funk3.jpg).
> 
> I can confirm that [828ba0c](https://github.com/mitmproxy/mitmproxy/commit/828ba0c2e79d8c54806a1c9eefb6007abc8aabc0) works for _both_ URLs.
> 
> Then [c0f62cc](https://github.com/mitmproxy/mitmproxy/commit/c0f62cc559ae4f0778d6079a2b31d2ce0309dda6) broke both.
> 
> And then HEAD (currently [0f96324](https://github.com/mitmproxy/mitmproxy/commit/0f963248872b3ae7c6d9b5065cd9ea79ae73f153)) only works for one of them.

I have tested this many times and found that the 413 error does not occur 100% of the time.
It's possible that it will be normal after a while, but these cases have one thing in common, all the servers that return exceptions are using Tengine (https://tengine.taobao.org/), which is a Nginx-based HTTP server.
www.taobao.com,www.bilibili.com uses Tengine, and you can see a lot of 413 code errors when you visit both sites.
Hopefully, this finding will help.
(I've also found a lot of 501 code errors that appear and disappear for no apparent reason, not sure if it's due to a change in the h2 module version.)

**Comment by Kriechi** [comment]
@Prinzhorn thanks for bisecting! 🎉 

I can't spot an obvious bug - but based on past experience I'm inclined to blame a faulty or non-RFC-compliant implementation in tengine or an outdated library on the server... We've seen lots of incompatibilities in the early days of HTTP/2. Some vendors even claimed "due to performance benefit we chose to break RFC compliance"...

I'll try to find some time and do another deep-dive, but unless anyone reports a similar problem with other webserver (most importantly the "real" nginx), I think we need to find a workaround for tengine specifically...

**Comment by Prinzhorn** [comment]
> @Prinzhorn thanks for bisecting! tada

@Kriechi that was mostly @bigfoxtail's work

> I can't spot an obvious bug - but based on past experience I'm inclined to blame a faulty or non-RFC-compliant implementation in tengine or an outdated library on the server... We've seen lots of incompatibilities in the early days of HTTP/2. Some vendors even claimed "due to performance benefit we chose to break RFC compliance"...
> 
> I'll try to find some time and do another deep-dive, but unless anyone reports a similar problem with other webserver (most importantly the "real" nginx), I think we need to find a workaround for tengine specifically...

Even if the server is not RFC compliant, the URLs work in Firefox, Chromium and Burp Suite. Obviously they have a lot more resources to make their stack compatible to almost anything. I just wanted to point that out. But like you said, if this doesn't become a more widespread issue we might just wait and see what happens.

On the other hand it _did_ work at some point.

**Comment by bigfoxtail** [comment]
I found another module httpx that uses http2 that works fine.
In the httpcore module used by httpx I found this part of the code.
https://github.com/encode/httpcore/blob/master/httpcore/_sync/http2.py
Line 286:
```
        # Send the request.
        seen_headers = set(key for key, value in headers)
        has_body = (
            b"content-length" in seen_headers or b"transfer-encoding" in seen_headers
        )

        self.send_headers(method, url, headers, has_body, timeout)
        if has_body:
            self.send_body(stream, timeout)
```
In c0f62cc559ae4f0778d6079a2b31d2ce0309dda6 you have modified some codes
```
-                end_stream=self.no_request_body,
```
and 
```
-        if not self.no_request_body:
-            self.connections[self.server_conn].safe_send_body(
-                self.raise_zombie,
-                self.server_stream_id,
-                chunks
-            )
```
Data was sent without determining the existence of the request body.
That seems to be the problem.

**Comment by Kriechi** [comment]
I've pushed a PR with a potential fix: #4239
Please let me know if this looks good and works for you. I also included a new test to better handle this case.

**Comment by Prinzhorn** [comment]
Can confirm that both URLs now work on master
